python/ray/data/dataset.py

```diff
@@ -2842,6 +2851,82 @@ def _experimental_lazy(self) -> "Dataset[T]":
         self._lazy = True
         return self
 
+    def is_out_of_band_serializable(self) -> bool:
+        """Whether this dataset is able to be out-of-band serialized, i.e. serialized
+        for use across different Ray clusters. Only datasets that read from lazy
+        datasources (i.e. via one of the ray.data.read_*() APIs) are out-of-band
+        serializable.
+        """
+        return self._plan.has_lazy_input()
+
+    @DeveloperAPI
+    def serialize_out_of_band(self) -> bytes:
+        """
+        Serialize the Dataset for out-of-band use, i.e. for use across different Ray
+        clusters. This method serializes the lineage of the Dataset operations. Note
+        that this will drop all computed data, and that everything will be recomputed
+        from scratch after deserialization.
+
+        Use ``Dataset.deserialize_out_of_band`` to deserialize the serialized bytes
+        into a Dataset.
+
+        Returns:
+            Serialized bytes.
+        """
+        if not self.is_out_of_band_serializable():
+            raise ValueError(
+                "Out-of-band serialization is only supported for Datasets created from "
+                "lazy datasources. I.e., out-of-band serialization is not "
```

Comment:
Agreed that "lazy" datasources isn't the best categorization. However, the datasources that we support don't necessarily need to be persistent storage systems, they only need to read data from a system external to the Ray cluster. E.g., if we had an Arrow Flight datasource that reads distributed in-memory data from an external cluster (Ray, Spark, or otherwise) that we know will exist when the experiment will be resumed, then that would be supported.

What do you think about changing it to something like this?
```python
"Lineage-based serialization is only supported for Datasets created from external data that we know will still exist at deserialization time, e.g. persistent cloud object stores or in-memory data from long-lived clusters."
```

Reply:
Sounds good -- reducing number of concepts our users need is great.